### ISSUE TITLE
Include SetAfterSaveBehavior as attribute [DatabaseGenerated] parameter

### ISSUE THREAD

**Issue by preira** [dotnet/efcore#31767]
At this point I have lots of minimal web api that only use attributes to describe the model, but, since EF doesn't allow naturally overriding default column values, I am facing the need to add extra complexity by including OnModelCreating to call .Metadata.SetAfterSaveBehavior(PropertySaveBehavior.Save);.

This should be as simple as passing a second argument to the attribute [DatabaseGenerated] but it isn't. So let's make it like that.

Regards,
João

**Comment by preira** [comment]
I am build historic an current data in the same table. I have a field that I want to be generated when the first insert for a given tracked entity happens and I want to control with the insert expression when I insert the next record for the same entity. E.g. `{id:1, snapshot:1, <other stuff>}; {id:1, snapshot:2}` ... when `insert (snapshot, ...) values (1, ...)`, `id` should be `++id`, when `insert (id, snapshot, ...) values (3, 2, ...)`, `id` should be `3`.

Postgres serial does just that, so I have a solution for some cases. In other cases I must default to older techniques like having a sequence table.

https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-serial/

I can also see this to be usefull with timestamps. Though one can always use "_top language_" defaults instead of column defaults.

**Comment by roji** [comment]
@preira I'm not clear on exactly what you're asking - it doesn't seem to be related to your previous question above (about an attribute for SetAfterSaveBehavior).

Are you looking for an auto-increment ID column, like PostgreSQL serial? If so, EF configures this automatically for you when it detects a numeric II property. For example, if you have a Blog entity type with an integer Id property, inserting instances of Blog will cause the Id to be auto-generated at the database side, very similar to a PostgreSQL serial.

Note that this mechanism - as well as PostgreSQL serial - performs the value generation in the database, and not at the client. In other words, the ID value is not included in the INSERT SQL which EF sends to the database (as in your comment above), but rather generated in the database and **returned** by the INSERT statement. Auto-increment keys must be generated in the database since that's the only way to ensure their uniqueness.

Finally, note that PostgreSQL serial has been largely deprecated in favor of IDENTITY columns, which are better. If you're using the EF PostgreSQL provider, it will automatically set up an IDENTITY column for numeric ID columns.

I suggest reading the EF docs, and especially the docs on [keys and value generation](https://learn.microsoft.com/en-us/ef/core/modeling/keys?tabs=data-annotations) for more info.

**Comment by preira** [comment]
The difference is that auto-increment ID column is always generated by database and you cannot override values in insert statements.

I don't see indication of deprecation of type serial for postgres, I can only find a don't use statement that I read has don't use if you don't know what you are doing.
https://wiki.postgresql.org/wiki/Don't_Do_This#Don.27t_use_serial

As said, I can always default to legacy pattern of having a sequences tables for the behaviour I intend.

Thanks for the challenging! 

You may close if it is not of value for the community.

**Comment by roji** [comment]
> The difference is that auto-increment ID column is always generated by database and you cannot override values in insert statements.

This isn't true (at least in PostgreSQL); a database-generated auto-increment ID - both the older serial and the newer IDENTITY columns - can be overridden by sending an explicit value in the INSERT VALUES. If you don't supply a value, it gets generated in the database, otherwise the provided value is used. For IDENTITY columns, that's precisely the difference between GENERATED BY DEFAULT AS IDENTITY and GENERATED ALWAYS AS IDENTITY - the latter does not allow overriding (although you can specify OVERRIDING SYSTEM VALUE on INSERT).

> I don't see indication of deprecation of type serial for postgres, I can only find a don't use statement that I read has don't use if you don't know what you are doing.

Note the following statement in that page which you linked to:

> For new applications, identity columns should be used instead.

Basically serial columns are the old way of doing auto-increment in PostgreSQL, identity columns are the new way.

> As said, I can always default to legacy pattern of having a sequences tables for the behaviour I intend.

Once again, there's absolutely no reason to do this - identity columns (and also serial) support insertion of explicit values from the client perfectly well. 

To summarize, you should be using PostgreSQL identity columns unless you have a very good reason not to (and I haven't seen one mentioned yet in this issue).

I'll go ahead and close this issue as there's nothing actionable here on the EF Core side, but if you have further questions on any of the above feel free to post back here and I'll do my best to explain.
